src/content/developers/docs/data-structures/patricia-merkle-trie/index.md

```diff
@@ -0,0 +1,217 @@
+```
+
+Here is the extended code for getting a node in the Merkle Patricia trie:
+
+```
+    def get_helper(node,path):
+        if path == []: return node
+        if node = '': return ''
+        curnode = rlp.decode(node if len(node) < 32 else db.get(node))
+        if len(curnode) == 2:
+            (k2, v2) = curnode
+            k2 = compact_decode(k2)
+            if k2 == path[:len(k2)]:
+                return get(v2, path[len(k2):])
+            else:
+                return ''
+        elif len(curnode) == 17:
+            return get_helper(curnode[path[0]],path[1:])
+
+    def get(node,path):
+        path2 = []
+        for i in range(len(path)):
+            path2.push(int(ord(path[i]) / 16))
+            path2.push(ord(path[i]) % 16)
+        path2.push(16)
+        return get_helper(node,path2)
+```
+
+### Example Trie {#example-trie}
+
+Suppose we want a trie containing four path/value pairs `('do', 'verb')`, `('dog', 'puppy')`, `('doge', 'coin')`, `('horse', 'stallion')`.
+
+First, we convert both paths and values to `bytes`. Below, actual byte representations for _paths_ are denoted by `<>`, although _values_ are still shown as strings, denoted by `''`, for easier comprehension (they, too, would actually be `bytes`):
+
+```
+    <64 6f> : 'verb'
+    <64 6f 67> : 'puppy'
+    <64 6f 67 65> : 'coin'
+    <68 6f 72 73 65> : 'stallion'
+```
+
+Now, we build such a trie with the following key/value pairs in the underlying DB:
+
+```
+    rootHash: [ <16>, hashA ]
+    hashA:    [ <>, <>, <>, <>, hashB, <>, <>, <>, [ <20 6f 72 73 65>, 'stallion' ], <>, <>, <>, <>, <>, <>, <>, <> ]
+    hashB:    [ <00 6f>, hashD ]
+    hashD:    [ <>, <>, <>, <>, <>, <>, hashE, <>, <>, <>, <>, <>, <>, <>, <>, <>, 'verb' ]
+    hashE:    [ <17>, [ <>, <>, <>, <>, <>, <>, [ <35>, 'coin' ], <>, <>, <>, <>, <>, <>, <>, <>, <>, 'puppy' ] ]
+```
+
+When one node is referenced inside another node, what is included is `H(rlp.encode(x))`, where `H(x) = sha3(x) if len(x) >= 32 else x` and `rlp.encode` is the [RLP](/fundamentals/rlp) encoding function.
+
+Note that when updating a trie, one needs to store the key/value pair `(sha3(x), x)` in a persistent lookup table _if_ the newly-created node has length >= 32. However, if the node is shorter than that, one does not need to store anything, since the function f(x) = x is reversible.
+
+## Tries in Ethereum {#tries-in-ethereum}
+
+All of the merkle tries in Ethereum's execution layer use a Merkle Patricia Trie.
+
+From a block header there are 3 roots from 3 of these tries.
+
+1.  stateRoot
+2.  transactionsRoot
+3.  receiptsRoot
+
+### State Trie {#state-trie}
+
+There is one global state trie, and it updates over time. In it, a `path` is always: `sha3(ethereumAddress)` and a `value` is always: `rlp(ethereumAccount)`. More specifically an ethereum `account` is a 4 item array of `[nonce,balance,storageRoot,codeHash]`. At this point it's worth noting that this `storageRoot` is the root of another patricia trie:
```

Comment:
It's more accurate to use `keccak256(ethereumAddress)` wherever `sha3` is mentioned in the Ethereum context since there's a tiny difference between the standardised `sha3` and the `keccak` draft proposed.

Reply:
ok, thanks - updated.